modules/renter/proto/negotiate.go

```diff
@@ -105,7 +104,7 @@ func verifyRecentRevision(conn net.Conn, contract modules.RenterContract) error
 	// check that revision number matches; if it does, do a more thorough
 	// check by comparing unlock hashes.
 	if lastRevision.NewRevisionNumber != contract.LastRevision.NewRevisionNumber {
-		return fmt.Errorf("our revision number (%v) does not match the host's (%v)", contract.LastRevision.NewRevisionNumber, lastRevision.NewRevisionNumber)
+		return &recentRevisionError{contract.LastRevision.NewRevisionNumber, lastRevision.NewRevisionNumber}
```

Comment:
If the contractor sees a recent revision error, it will start over (i.e. close the connection and make a new RPC) using the saved revision.

One failing of the current implementation is that it does not reconstruct the revision transaction in the event of a desync. However, this transaction is only stored so that it might be submitted to the blockchain later, and as you pointed out, this is rarely in the renter's interest anyway. Also, the revision transaction will be correctly updated as soon as the next revision completes.